### ISSUE TITLE
ParserError is missing end position

### ISSUE THREAD

**Issue by tscpp** [meriyah/meriyah#156]
I need the end position of any parser error for my linter, but this is not an available property in the ParserError class.

**Comment by 3cp** [comment]
@KFlash this is probably going to be a breaking change.
Changing `loc: {line, column}` to `loc: {start: {line, column}, end: {line, column}}`.

**Comment by KFlash** [comment]
What other parsers do ?

**Comment by 3cp** [comment]
I see your point. It seems babel/parser also has the `loc: {line, column}`.

```
{
  "loc": {
    "line": 1,
    "column": 9
  },
  "pos": 9
}
SyntaxError: Unterminated regular expression (1:9)
```

@tscpp what was your previous parser?

**Comment by KFlash** [comment]
@3cp I will ask him. I know his language

@tscpp  Kan du utarbeta lite mer om vad du menar? Alla "parsers" gör samma sak. Och i Meriyah är detta baserat på ECMA-standarden.  Vad är skillnaden mellan Meriyah och den tidigare koden du använde? Kan det räcka att lägga till en extra "property"? Kommer det att hjälpa dig?

**Comment by tscpp** [comment]
The projected used Acorn before. I switched to Meriyah because of the performance difference and the support for ES2020. We didn't log the syntax error location before, we logged the error to the whole line. I got suggested to log the syntax error meanwhile I was working with the pull request. We want support for end positions but it was not available, so I created this issue.

If this is possible, my suggestion would be to add a property called `range` or `start` and `end` in ParserError before the release of a new master version. Also, could ParserError or an interface of ParserError be exported in the main module?

**Comment by KFlash** [comment]
@3cp It shouldn't be too hard to add extra property? See [this code](https://github.com/meriyah/meriyah/blob/master/src/errors.ts#L350). Loc already exist. Shouldn't affect the rest of the AST nodes.

@tscpp Sounds okay? I'm not sure what language you prefer? Är detta en okej lösning på problemet? För att lägga till det här?
Jag arbetade 11 år i grannlandet. Strax över gränsen talar också norska.

**Comment by tscpp** [comment]
Jag föredrar engelska så att andra kan läsa mina kommentarer 😁

----

I'm making a pull request adding the property `end` to `ParserError`.

**Comment by tscpp** [comment]
Should `'[' + loc.start.line + ':' + loc.start.column + ']: '` be removed from the message. I'm finding this quite unseccesary. It should be added by the client.

**Comment by KFlash** [comment]
I don't think it's an standard on this. All Parsers report this differently. Try to look what others do with AST explorer

**Comment by tscpp** [comment]
`ParserState` contains end. What position is that?

**Comment by KFlash** [comment]
See `finishNode` in common. TS file. Add start and if that solve the issue

**Comment by tscpp** [comment]
Should I use `ParserState`'s `line` and `column` for that. `finishNode` has parameters `context: Context` and `node: any` which I don't know where to find.

**Comment by tscpp** [comment]
> I don't think it's an standard on this. All Parsers report this differently. Try to look what others do with AST explorer

Acorn uses SyntaxError which doesn't report the start position.

**Comment by tscpp** [comment]
Because the variable node is not defined in every function in `parser.ts`, I need to define the variable node at 194 places in the code. I added the parameters `context` and `node` to the function `report`, and it seems like `node` is not defined anywhere. I also looked at the references of `finishNode`, and all of the expressions calling the function define node with properties `type` and `body`. I have `body`, but the problem is that I don't know `type`, I could define all node variables but it seems like a bit more work than I'm currently able to do. Do you still want to rest of the pull request?

**Comment by KFlash** [comment]
cc @3cp Could you help out with this?

**Comment by 3cp** [comment]
It's hard to follow without looking at the code. You can make a draft PR, doesn't need to be fully functional, but provide a base for us to discuss.

**Comment by tscpp** [comment]
@3cp #157
